apps/blockchain/lib/blockchain/blocktree.ex

```diff
@@ -77,7 +77,15 @@ defmodule Blockchain.Blocktree do
          do: block,
          else: best_block
 
-    %{blocktree | best_block: new_best_block}
+    # Make sure block is stored with a block hash
+    new_best_block_with_hash =
+      if new_best_block.block_hash == nil do
+        %{new_best_block | block_hash: Block.hash(new_best_block)}
```

Comment:
I think this may cause multiple calculations of hash for the same block. Because it's calculated when the block is stored

Reply:
It's a bit weird, it's really meant to be a write-through cache, as in, don't calculate it until you need it, but then you should store it in the block to not have to calculate it again. If, downstream, the code reads block hash, it should really check to see if it was already calculated here.

Reply:
Also, we should probably just add a function in `Block` that is `Block.with_hash` that just handles this logic.